Paron

skjohansen opened the pull request.

MarksPG committed:
Lade till en .gitignore-fil

CarljohanDanling committed:
no message

JohanWWW committed:
Project code init

TomasVaro committed:
UserStories

TomasVaro committed:
Merge branch 'Paron' of https://github.com/PGBFDH18/ludo-console-game-projektgrupp_paron into Paron

TomasVaro committed:
Create test.txt

TomasVaro committed:
Merge branch 'Paron' of https://github.com/PGBFDH18/ludo-console-game-projektgrupp_paron into Paron

JohanWWW committed:
Ändrade till .net core och lagt till bibliotek

JohanWWW committed:
Merge branch 'Paron' of https://github.com/PGBFDH18/ludo-console-game-projektgrupp_paron into Paron

TomasVaro committed:
Merge branch 'Paron' of https://github.com/PGBFDH18/ludo-console-game-projektgrupp_paron into Paron

JohanWWW committed:
moved source code to src

MarksPG committed:
no message

MarksPG committed:
Merge branch 'Paron' of https://github.com/PGBFDH18/ludo-console-game-projektgrupp_paron into Paron

TomasVaro committed:
Merge branch 'Paron' of https://github.com/PGBFDH18/ludo-console-game-projektgrupp_paron into Paron

JohanWWW committed:
Lagt till kod. (Startat)

JohanWWW committed:
Flyttat Piece till GameEngine. Gjort ändringar.

JohanWWW committed:
testing testing testing

MarksPG committed:
Testar lite med Dokumentationsfilen

MarksPG committed:
Tog bort en readme-fil enligt anvisning och flyttade md-filen till mappen docs

MarksPG committed:
no message

MarksPG committed:
Ytterligare lite trix med dokumentationen

JohanWWW committed:
Klart:
* Player
* Piece
* Dice

Fortsätt på:
* Utskrift på pjäs
* Spellogik

JohanWWW committed:
Merge branch 'Paron' of https://github.com/PGBFDH18/ludo-console-game-projektgrupp_paron into Paron

JohanWWW committed:
Lagt till punkt i User stories.

TomasVaro committed:
Ändrat i UserStories

JohanWWW committed:
#Mindre korrigeringar
- Lagt till metoder i Piece
- Städat

JohanWWW committed:
Ändrat

JohanWWW committed:
# GameEngine
- Konstruktorn i Game behöver inte första deklarationen då den omdeklarareras direkt efter.
- Lagt till klassmetoden IsOnOpponentPosition(...) i Game som kontrollerar om en pjäs står på en annan pjäs position.
- Player har en ToString() som returnerar strängen "Spelare X" där X är siffran på spelaren. Värdet returneras när instansen placeras i en sträng.
- Konstruktorn i Piece tar nu emot correctionFactor som argument. Korrigeringsfaktorn tilldelas i samband med att instansen skapas.
- Rättat till metoden GetAbsolutePosition(). När värdet på GetAbsolutePosition() blir 41 "återställs" det till 1.
- Home() återställer pjäsens position till 0.
- IsHome() kontrollerar om pjäsen står på position 0.
- InGoal() kontrollerar om pjäsen har nått mål.
- Piece har metod ToString() som automatiskt formaterar pjäsens färg och siffra till en sträng när instansen placeras i en sträng.
- Metod Roll() är nu statisk så att den inte behöver instantieras för att användas.

# LudoGame
- If-sats som kontrollerar att antalet spelare ligger mellan 2-4.
- For-loopen delar ut korrigeringsfaktorer till spelare med correctionFactor.
- Lagt till metoden PlayerTurn(...). Metoden motsvarar en "spelaromgång" och anropas därför för varje spelare.
- Mycket av besluten som görs ligger i metoden PlayerTurn(...).

# Det som behöver fixas
- Presentationen (/ användarvänligheten).

JohanWWW committed:
* Pjäsernas globala position visas nu tillsammans med den lokala positionen i utskriften.
* Städade och skrubbade.
* Flyttade metoden IsOnOpponentPosition(...) som tidigare låg i Game, till Piece där den verkar bättre höra hemma. Metoden finns även som instansmetod om man vill ha lite saker att välja på :)

Giss47 commented on src/LudoGame/GameEngine/Class1.cs:
```diff
@@ -52,12 +52,12 @@ public class Piece
         public int correctionFactor { get; set; }
```
Can you please explain what you mean by correctionFactor

JohanWWW replied:
Varje pjäs har en "lokal" position (antal steg som pjäsen har gått) och en "global" position (vilken ruta på brädan pjäsen står på). De första fyra pjäserna som skapas för en spelare får correctionFactor = 0. Därefter ökas värdet med 10 för varje fjärde pjäs (0, 10, 20, 30). Om vi antar att en pjäs har correctionFactor = 20 och den har gått 16 steg kan vi räkna ut den absoluta positionen (var på brädan den är) genom att ta position + correctionFactor = 20 + 16 = 36. Den befinner sig alltså på ruta 36. Genom att räkna ut den absoluta/globala positionen kan vi jämföra två pjäser med varandra för att avgöra om de står på samma ruta.

Sammanfattat är correctionFactor en "konstant" som adderas med pjäsens lokala position för att få fram dess globala position.

skjohansen replied:
Skulle vara bra om denna beskrivning på något sätt är att finna i eran docs-mapp.

CarljohanDanling replied:
Resolved.

skjohansen commented on src/LudoGame/GameEngine/Class1.cs:
```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace GameEngine
```
There is nothing wrong in you approach, but I would recommend to  only have one class per code file. So you should have one file called Game.cs containing the Game class, one called Player.cs containing the Player class, etc.

CarljohanDanling replied:
This is now changed. All classes in GameEngine are now in separate files.

skjohansen commented on src/LudoGame/GameEngine/Class1.cs:
```diff
@@ -0,0 +1,90 @@
+        // Deltagarna ligger här!
+        public List<Player> Players { get; set; }
+
+        public Game(List<Player> players)
```
Why have a constructor which sets the public property players?

skjohansen commented on src/LudoGame/GameEngine/Class1.cs:
```diff
@@ -0,0 +1,90 @@
+
+    public class Piece
+    {
+        public PieceColor Color { get; set; }
```
Are there any specific reason that this and the two following lines are 1) properties 2) public?

skjohansen commented on src/LudoGame/GameEngine/Class1.cs:
```diff
@@ -0,0 +1,90 @@
+        public bool InGoal() => position >= 45;
+
+        // Ersatts av instansvariabeln. Överge?
+        public static bool IsOnOpponentPosition(Piece myPiece, Piece opponentPiece) =>
```
Where is this used?

skjohansen commented on src/LudoGame/GameEngine/Class1.cs:
```diff
@@ -0,0 +1,90 @@
+            GetAbsolutePosition() == opponentPiece.GetAbsolutePosition() && this != opponentPiece;
+
+        // Exempelresultat: 'R3' om pjäsen är Röd och har siffra 3.
+        public override string ToString() => $"{Color.ToString()[0]}{Number}";
```
👍 Nice!

skjohansen commented on src/LudoGame/LudoGame/Program.cs:
```diff
@@ -0,0 +1,178 @@
+                var index = int.Parse(Console.ReadLine());
+                Console.Clear();
+
+                // Tilldela startpostitioner till pjäser.                
```
Seems like game logic.

skjohansen commented on src/LudoGame/LudoGame/Program.cs:
```diff
@@ -0,0 +1,178 @@
+            Console.WriteLine();
+            // Spelaren väljer här vilken pjäs som ska flyttas från utifrån listan.
+            Console.Write("Mata in index: ");
+            int choice = int.Parse(Console.ReadLine()); // Behöver felhantering.
```
"Behöver felhantering." - Just do it 😁

skjohansen replied:
```suggestion
            bool parseSucceded = int.TryParse(Console.ReadLine(), int out choice);
            if(!parseSucceded){
                // Do something
            }
```

skjohansen commented on src/LudoGame/LudoGame/Program.cs:
```diff
@@ -0,0 +1,178 @@
+            Piece piece = player.Pieces[choice - 1];    // Behöver felhantering.
+
+            // Kontrollerar om pjäsen är inne i boet och tärningslagen är 1 eller 6.
+            if (piece.IsHome() && steps == 1 || steps == 6)
```
Seems like game logic

skjohansen reviewed (commented):
Nice working game
It seems that not all parts of the logic is located in the game logic.. maybe something to look into.
Split the Class1 into several files, except that quite nice code!

CarljohanDanling committed:
Alla klasser i GameEngine ligger nu i separata filer. Enum flyttades in i Piece-klassen. Ändrade anrop i Program för PieceColor.

Commit:
#Dice
- Added modifier sealed to prevent inheritance.
- Added XML documentation comments.

#Piece
- Added XML documentation comments.
- Bug fix: GetAbsolutePosition() now returns 0 when position is 0
- Replaced conditional operator syntax with an IF ELSE statement due to the complexity in which the additional condition (above) provided.

#Player
- Added XML documentation comments.

#Program
- Renamed ListPieces() to ListAllPieces().
- Added method ListPlayerPieces(in: Player) which prints list of a player's pieces in which the player can choose from.

Commit:
Revert "#Dice"

This reverts commit fddfcbdc917362544adddb3eeb83f3a50ca602b3.

Commit:
Revert "Revert "#Dice""

This reverts commit f6d48caa8f97d2016a4812184aebe38cb3bd44c5.

Commit:
Revert "Revert "Revert "#Dice"""

This reverts commit f74257e716174dfab2cc291603bb1354411edfc2.

Commit:
misstag i commit fixad

#Dice
- Added modifier sealed to prevent inheritance.
- Added XML documentation comments.

#Piece
- Added XML documentation comments.
- Bug fix: GetAbsolutePosition() now returns 0 when position is 0
- Replaced conditional operator syntax with an IF ELSE statement due to the complexity in which the additional condition (above) provided.

#Player
- Added XML documentation comments.

#Program
- Renamed ListPieces() to ListAllPieces().
- Added method ListPlayerPieces(in: Player) which prints list of a player's pieces in which the player can choose from.

Commit:
Fel ändamål i förra pushen. Syftet var att inte göra det möjligt att skapa en instans av klassen. Static förhindrar instantiering (även arv) av Dice.

MarksPG committed:
Lade in lite extra text - correctionFactor - i Userstories. Kan ses som preliminärt.

MarksPG committed:
Merge branch 'Paron' of https://github.com/PGBFDH18/ludo-console-game-projektgrupp_paron into Paron

JohanWWW committed:
* Lagt till SQL-bibliotek.
* Lagt till Save och Load funktionalitet (Database och Storage av IProvider).

JohanWWW committed:
* Ändrade namn på IProvider till ett bättre beskrivande namn: ILoadSave.
* Ändrade namn på Storage till Drive och är alltså klassen som ska ladda och spara spel till och från hårddisken.
* Ändrade namnet på propertyn Database i Game till LoadSaveMethod då det inte bara är typen Database som den lagra.

AnorZaken reviewed (changes requested):
Ser väl bra ut, har du tänkt fylla ut NotImplemented delarna nu, eller ska det bara approve'as och fixas senare?
(Seg som *** på att kolla min mail, sorry)

AnorZaken reviewed (changes requested):
LoadSaveMethod är ju inte en metod så borde heta något annat (hade passat om det var en delegate/Action<>/Func<>).
Är väl inte fel att säga att det är en komponent som Game använder, kanske LoadSaveComponent?
Låter korrekt och man förstår direkt vad den är till för, fast kanske fortfarande lite meh/långt.... kanske ska man lite mer fritt döpa den till SaveSlot?
Namngivning är svårt! :P 
(Annat förslag är t.ex, att istället döpa den till bara LoadSave och istället / eventuellt ändra ILoadSave till ILoadSaveComponent - men 'method' passar inte eftersom det har en särskild betydelse som kan förvirra andra programmerare.)
